llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,527 @@
+void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
+                               FunctionToAspectsMapTy &AspectsMap,
+                               SmallPtrSet<Function *, 16> &Visited) {
+  auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+
+  AspectToFunctionLinkMapTy LocalAspects;
+  for (auto Edge : It->second) {
+    Function *Callee = Edge.first;
+    if (Visited.insert(Callee).second)
+      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+
+    auto &CalleeAspects = AspectsMap[Callee];
+    for (auto AspectIt : CalleeAspects) {
+      LocalAspects[AspectIt.first] = FunctionLinkTy{Callee, Edge.second};
+    }
+  }
+
+  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+/// Processes function's instructions. It analyzes aspect usages with debug
+/// information and builds a call graph.
+void processFunctionInstructions(Function &F,
+                                 FunctionToAspectsMapTy &FunctionToAspects,
+                                 TypeToAspectsMapTy &TypesWithAspects,
+                                 CallGraphTy &CG) {
+  auto &AspectToFunctionLinkMap = FunctionToAspects[&F];
+  for (Instruction &I : instructions(F)) {
+    AspectsSetTy Aspects = getAspectsUsedByInstruction(I, TypesWithAspects);
+    for (int Aspect : Aspects) {
+      FunctionToAspects[&F][Aspect] = FunctionLinkTy{nullptr, &I.getDebugLoc()};
+    }
+
+    if (auto *DBI = dyn_cast<DbgVariableIntrinsic>(&I)) {
+      // Handle a group of llvm.dbg.{addr,declare,value} intrinsics.
+      // Example:
+      // %tmp = alloca %Struct
+      // call void @llvm.dbg.declare(metadata %Struct* %tmp, metadata !1,
+      //                             metadata !DIExpression), !dbg !2
+      //
+      // Here we extract the first intrinsic argument, then extract
+      // %Struct type, then analyze known aspects of %Struct and update
+      // corresponding aspect's records with debug information !dbg !2.
+      Type *T = DBI->getVariableLocationOp(0)->getType();
+      Aspects = getAspectsFromType(T, TypesWithAspects);
+      for (int Aspect : Aspects) {
+        auto It = AspectToFunctionLinkMap.find(Aspect);
+        if (It == AspectToFunctionLinkMap.end())
+          continue;
+
+        It->second.DL = &I.getDebugLoc();
+      }
+
+      continue;
+    }
+
+    if (auto *CI = dyn_cast<CallInst>(&I)) {
+      if (!CI->isIndirectCall() && CI->getCalledFunction())
+        CG[&F].try_emplace(CI->getCalledFunction(), &CI->getDebugLoc());
+    }
+  }
+}
+
+/// Returns a map of functions with corresponding used aspects.
+FunctionToAspectsMapTy
+buildFunctionsToAspectsMap(Module &M, TypeToAspectsMapTy &TypesWithAspects) {
+  FunctionToAspectsMapTy FunctionToAspects;
+  CallGraphTy CG;
+  std::vector<Function *> Kernels;
+  for (Function &F : M.functions()) {
+    auto CC = F.getCallingConv();
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (CC == CallingConv::SPIR_KERNEL)
```

Comment:
Module "roots" (or "entry points") for the aspect propagation should also include SYCL_EXTERNAL functions. In fact, there can be modules w/o kernels at all.
See [example ](https://github.com/intel/llvm/blob/4072557b121d0fb322a5e167cc8905209f21ab3e/llvm/tools/sycl-post-link/sycl-post-link.cpp#L304) from post-link how entry points can be determined.

Reply:
Fixed and added a test in b558a2d1f3178d14ab74658750d00d8d1137dddb